src/translator/cache.cpp

```diff
@@ -0,0 +1,157 @@
+#include "translator/cache.h"
+
+#include <cstdlib>
+
+namespace marian {
+namespace bergamot {
+
+#ifndef WASM_COMPATIBLE_SOURCE
+
+ThreadSafeL4Cache::ThreadSafeL4Cache(Ptr<Options> options)
+    : epochManagerConfig_(options->get<size_t>("cache-ebr-queue-size"),
+                          std::chrono::milliseconds(options->get<size_t>("cache-ebr-interval")),
+                          options->get<size_t>("cache-ebr-num-queues")),
+      cacheConfig_(options->get<size_t>("cache-size") * 1024 * 1024,
+                   std::chrono::seconds(options->get<size_t>("cache-time-to-live")),
+                   options->get<bool>("cache-remove-expired")),
+      service_(epochManagerConfig_),
+      context_(service_.GetContext()) {
+  // There is only a single cache we use. However, L4 construction API given by example demands we give it a string
+  // identifier.
+  const std::string cacheIdentifier = "global-cache";
+  size_t numBuckets = options->get<size_t>("cache-buckets");
+  hashTableIndex_ = service_.AddHashTable(L4::HashTableConfig(
```

Comment:
What is the role of the cache buckets here? is there a reason to have more than one? Are there performance implications?

Reply:
I assume all threads go to the same bucket, otherwise you would have a lot of cache misses.

Reply:
I'll have to look closer. My guess would be that buckets contain granularity of locking coarse/fine-grained to reduce contention. I've seen mutexes per bucket.. so.

Reply:
After reading the code more carefully, buckets are a cache hyperparameter that needs to be benchmarked. More buckets might mean better performance in multithreaded case.

Reply:
Buckets don't really matter all that much in testing